subprojects/gdk_rust/gdk_electrum/src/lib.rs

```diff
@@ -0,0 +1,1212 @@
+                let max = result
+                    .iter()
+                    .enumerate()
+                    .filter(|(_, v)| !v.is_empty())
+                    .map(|(i, _)| i as u32)
+                    .max();
+                if let Some(max) = max {
+                    last_used[i as usize] = max + batch_count * BATCH_SIZE;
+                };
+
+                let flattened: Vec<GetHistoryRes> = result.into_iter().flatten().collect();
+                trace!("{}/batch({}) {:?}", i, batch_count, flattened.len());
+
+                if flattened.is_empty() {
+                    break;
+                }
+
+                for el in flattened {
+                    if el.height >= 0 {
+                        heights_set.insert(el.height as u32);
+                        txid_height.insert(el.tx_hash, el.height as u32);
+                    }
+                    history_txs_id.insert(el.tx_hash);
+                }
+
+                batch_count += 1;
+            }
+        }
+
+        self.db.insert_index(Index::External, last_used[Index::External as usize])?;
+        self.db.insert_index(Index::Internal, last_used[Index::Internal as usize])?;
+        trace!("last_used: {:?}", last_used,);
+
+        let new_txs = self.sync_txs(&history_txs_id)?;
+        self.sync_headers(&heights_set)?;
+        self.sync_height(&txid_height)?;
+
+        if new_txs {
+            self.db.flush()?;
+        }
+
+        trace!("elapsed {}", start.elapsed().as_millis());
+
+        Ok(new_txs)
+    }
+
+    fn sync_headers(&mut self, heights_set: &HashSet<u32>) -> Result<(), Error> {
+        let heights_in_db = self.db.get_only_heights()?;
+        let heights_to_download: Vec<u32> =
+            heights_set.difference(&heights_in_db).cloned().collect();
+        if !heights_to_download.is_empty() {
+            let headers_bytes_downloaded =
+                self.client.batch_block_header_raw(heights_to_download.clone())?;
+            let mut headers_downloaded: Vec<BEBlockHeader> = vec![];
+            for vec in headers_bytes_downloaded {
+                headers_downloaded.push(BEBlockHeader::deserialize(&vec, self.network.id())?);
+            }
+
+            for (header, height) in headers_downloaded.iter().zip(heights_to_download.iter()) {
+                self.db.insert_header(*height, header)?;
+            }
+            info!("headers_downloaded {:?}", headers_downloaded);
+        }
+
+        Ok(())
+    }
+
+    fn sync_height(&self, txid_height: &HashMap<Txid, u32>) -> Result<(), Error> {
+        // sync heights, which are my txs
+        for (txid, height) in txid_height.iter() {
+            self.db.insert_height(txid, *height)?; // adding new, but also updating reorged tx
+        }
+        for txid_db in self.db.get_only_txids()?.iter() {
+            if txid_height.get(txid_db).is_none() {
+                self.db.remove_height(txid_db)?; // something in the db is not in live list (rbf), removing
```

Comment:
Is removing the height associated with the tx sufficient to have the tx ignored everywhere where it needs to be ignored?

For example, will `get_all_spent_and_txs` know not to include spends made by rbfed/conflicted transactions into the `spent` HashSet?

(It is probably better to keep some record of the tx and not remove it entirely though. Ideally, I think that they should also be displayed to the user somewhere in the GUI, at least for incoming txs that ended up not paying the user's address.)

Reply:
the Txid, Height list helps to build the wallet txs, dealing with the "live" situation and dealing with reorgs and rbfs. Transaction that existed at any point in time aren't really deleted (even though maybe not "linked")

> For example, will get_all_spent_and_txs know not to include spends made by rbfed/conflicted transactions into the spent HashSet?

if spent contains something that doesn't exist anymore, I don't think it's a problem

Reply:
> if spent contains something that doesn't exist anymore, I don't think it's a problem

Wouldn't it mean that usable unspent coins could be incorrectly excluded from coin selection?

Reply:
> Wouldn't it mean that usable unspent coins could be incorrectly excluded from coin selection?

No, it would mean you haven't unusable unspent in your utxo

Maybe I am not getting the exact situation, could you make a step by step example (we may want to turn into a test) ?

Reply:
I'm thinking of a scenario where transaction `T1` spends the prevout `O1`, then gets double-spent by a transaction that doesn't spend `O1` (but does conflict with `T1` because its spends another one of its inputs), making `O1` unspent again.

Will the coin selection use `O1` in that case?

Reply:
I think we definitely should have some more tests for reorg scenarios, there's a lot of trickery around that...

Reply:
That's should be handled if electrum API doesn't return T1 anymore for prevout `O1 ` script.
But yes more tests are always a good thing